File: python/examples/02_k_matrix_anatomy.py
#!/usr/bin/env python3
"""Example 2 - what each entry of K actually does.

Run:  uv run python/examples/02_k_matrix_anatomy.py [--save out.png]

Renders the same 3D scene four times, changing exactly one thing each time, and
prints the field of view alongside.  The point to internalise: K contains no
rotation and no translation.  It cannot move the camera.  It only decides how
the already-projected ray lands on the sensor grid.
"""

import argparse
import os
import sys

import numpy as np

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from camintrinsics import fov_deg, make_K                       # noqa: E402
from camintrinsics.renderer import (draw_crosshair, draw_text_block,  # noqa: E402
                                    hstack_labeled, orbit_pose, render,
                                    save_image)
from camintrinsics.scene import default_scene                   # noqa: E402

W, H = 420, 320
BASE_F = 320.0


def shot(K, caption):
    img = np.full((H, W, 3), (24, 24, 28), np.uint8)
    render(img, default_scene(), orbit_pose([0, 0, 4.0], 4.4, 14.0, 10.0), K, None)
    draw_crosshair(img, K)
    hf, vf, _ = fov_deg(K, W, H)
    draw_text_block(img, [f"fx {K[0,0]:.0f}  fy {K[1,1]:.0f}",
                          f"cx {K[0,2]:.0f}  cy {K[1,2]:.0f}  skew {K[0,1]:.0f}",
                          f"FOV {hf:.1f} x {vf:.1f} deg"],
                    org=(10, 20), scale=0.42, line_h=16)
    return img, caption


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--save", default=None)
    args = ap.parse_args()

    variants = [
        (make_K(BASE_F, BASE_F, W / 2, H / 2), "baseline"),
        (make_K(BASE_F * 2, BASE_F * 2, W / 2, H / 2), "fx,fy x2  ->  zoom in, FOV halves"),
        (make_K(BASE_F, BASE_F * 1.7, W / 2, H / 2), "fy x1.7  ->  non-square pixels"),
        (make_K(BASE_F, BASE_F, W / 2 - 120, H / 2 + 60), "cx,cy moved -> image shifts, camera does not"),
    ]
    tiles = [shot(K, c) for K, c in variants]
    row1 = hstack_labeled([tiles[0][0], tiles[1][0]], [tiles[0][1], tiles[1][1]])
    row2 = hstack_labeled([tiles[2][0], tiles[3][0]], [tiles[2][1], tiles[3][1]])
    grid = np.vstack([row1, row2])

    print(__doc__)
    print(f"{'variant':46s} {'hFOV':>7s} {'vFOV':>7s} {'dFOV':>7s}")
    for K, cap in variants:
        hf, vf, df = fov_deg(K, W, H)
        print(f"{cap:46s} {hf:7.2f} {vf:7.2f} {df:7.2f}")

    print("\nthings worth noticing")
    print("  * fx and fy are focal lengths measured IN PIXELS, so they change")
    print("    when you resize the image even though the lens did not change.")
    print("  * fx != fy only means the pixels are not square (or someone")
    print("    stretched the image). Modern sensors give fx/fy within ~1%.")
    print("  * moving cx,cy slides the image; it does NOT rotate the camera.")
    print("    A rotation changes what is visible at infinity, a cx shift does not.")
    print("  * skew is the shear between the sensor axes. It is 0 for every")
    print("    digital sensor you will meet, and cv2.projectPoints ignores it")
    print("    entirely -- see docs/course/02_the_K_matrix.md.")

    if args.save:
        if not save_image(args.save, grid):
            return 1
    else:
        print("\n(pass --save out.png to write the figure)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: cpp/examples/02_k_matrix_anatomy.cpp
// Example 2 - what each entry of K actually does.
//
//   ./build/bin/02_k_matrix_anatomy [out.png]
//
// Renders the same 3D scene four times, changing exactly one thing each time.
// The point to internalise: K contains no rotation and no translation. It
// cannot move the camera. It only decides how the already-projected ray lands
// on the sensor grid.

#include <cstdio>
#include <string>
#include <vector>

#include "camintrinsics/renderer.hpp"
#include "camintrinsics/util.hpp"
#include "camintrinsics/scene.hpp"

namespace {
constexpr int W = 420, H = 320;
constexpr double kBaseF = 320.0;

cv::Mat shot(const ci::Mat33& K) {
  cv::Mat img(H, W, CV_8UC3, cv::Scalar(24, 24, 28));
  ci::render(img, ci::defaultScene(), ci::orbitPose({0, 0, 4.0}, 4.4, 14.0, 10.0),
             K, ci::makeD());
  ci::drawCrosshair(img, K);
  const cv::Vec3d f = ci::fovDeg(K, W, H);
  ci::drawTextBlock(img,
                    {ci::fmt("fx %.0f  fy %.0f", K(0, 0), K(1, 1)),
                     ci::fmt("cx %.0f  cy %.0f  skew %.0f", K(0, 2), K(1, 2),
                             K(0, 1)),
                     ci::fmt("FOV %.1f x %.1f deg", f[0], f[1])},
                    {10, 20}, 0.42, ci::colors::kWhite, 16);
  return img;
}
}  // namespace

int main(int argc, char** argv) {
  struct Variant { ci::Mat33 K; const char* caption; };
  const std::vector<Variant> variants = {
      {ci::makeK(kBaseF, kBaseF, W / 2.0, H / 2.0), "baseline"},
      {ci::makeK(kBaseF * 2, kBaseF * 2, W / 2.0, H / 2.0),
       "fx,fy x2  ->  zoom in, FOV halves"},
      {ci::makeK(kBaseF, kBaseF * 1.7, W / 2.0, H / 2.0),
       "fy x1.7  ->  non-square pixels"},
      {ci::makeK(kBaseF, kBaseF, W / 2.0 - 120, H / 2.0 + 60),
       "cx,cy moved -> image shifts, camera does not"}};

  std::printf("%-46s %7s %7s %7s\n", "variant", "hFOV", "vFOV", "dFOV");
  for (const Variant& v : variants) {
    const cv::Vec3d f = ci::fovDeg(v.K, W, H);
    std::printf("%-46s %7.2f %7.2f %7.2f\n", v.caption, f[0], f[1], f[2]);
  }

  std::printf("\nthings worth noticing\n");
  std::printf("  * fx and fy are focal lengths measured IN PIXELS, so they "
              "change\n    when you resize the image even though the lens did "
              "not change.\n");
  std::printf("  * fx != fy only means the pixels are not square (or someone\n"
              "    stretched the image). Modern sensors give fx/fy within ~1%%.\n");
  std::printf("  * moving cx,cy slides the image; it does NOT rotate the "
              "camera.\n    A rotation changes what is visible at infinity, a "
              "cx shift does not.\n");
  std::printf("  * skew is the shear between the sensor axes. It is 0 for "
              "every\n    digital sensor you will meet, and cv::projectPoints "
              "ignores it\n    entirely -- see docs/course/02_the_K_matrix.md.\n");

  if (argc > 1) {
    const cv::Mat row1 = ci::hstackLabeled({shot(variants[0].K), shot(variants[1].K)},
                                           {variants[0].caption, variants[1].caption});
    const cv::Mat row2 = ci::hstackLabeled({shot(variants[2].K), shot(variants[3].K)},
                                           {variants[2].caption, variants[3].caption});
    if (!ci::writeImage(argv[1], ci::vstack({row1, row2}))) return 1;
  } else {
    std::printf("\n(pass an output path to write the figure)\n");
  }
  return 0;
}
